flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cluster/KubernetesClusterDescriptor.java

```diff
@@ -0,0 +1,153 @@
+
+	private static final String CLUSTER_ID_PREFIX = "flink-session-cluster-";
+
+	private static final String CLUSTER_DESCRIPTION = "Kubernetes cluster";
+
+	private FlinkKubernetesOptions options;
+
+	private KubernetesClient client;
+
+	public KubernetesClusterDescriptor(@Nonnull FlinkKubernetesOptions options, @Nonnull KubernetesClient client) {
+		this.options = options;
+		this.client = client;
+	}
+
+	private String generateClusterId() {
+		return CLUSTER_ID_PREFIX + UUID.randomUUID();
+	}
+
+	@Override
+	public String getClusterDescription() {
+		return CLUSTER_DESCRIPTION;
+	}
+
+	private ClusterClient<String> createClusterEndpoint(Endpoint clusterEndpoint, String clusterId) throws Exception {
+
+		Configuration configuration = new Configuration(this.options.getConfiguration());
+		configuration.setString(JobManagerOptions.ADDRESS, clusterEndpoint.getAddress());
+		configuration.setInteger(JobManagerOptions.PORT, clusterEndpoint.getPort());
+		return new RestClusterClient<>(configuration, clusterId);
+	}
+
+	@Override
+	public ClusterClient<String> retrieve(String clusterId) throws ClusterRetrieveException {
+		try {
+			Endpoint clusterEndpoint = this.client.getResetEndpoint(clusterId);
+			return this.createClusterEndpoint(clusterEndpoint, clusterId);
+		} catch (Exception e) {
+			this.client.logException(e);
+			throw new ClusterRetrieveException("Could not create the RestClusterClient.", e);
+		}
+	}
+
+	@Override
+	public ClusterClient<String> deploySessionCluster(ClusterSpecification clusterSpecification)
+		throws ClusterDeploymentException {
+
+		String clusterId = this.generateClusterId();
+
+		//TODO: add arguments
+		final List<String> args = Arrays.asList();
+
+		return this.deployClusterInternal(clusterId, args);
+	}
+
+	@Override
+	public ClusterClient<String> deployJobCluster(ClusterSpecification clusterSpecification, JobGraph jobGraph, boolean detached) {
+		throw new NotImplementedException();
+	}
+
+	@Nonnull
+	private ClusterClient<String> deployClusterInternal(String clusterId, List<String> args) throws ClusterDeploymentException {
+		try {
+			Endpoint clusterEndpoint = this.client.createClusterService();
+			this.client.createClusterPod();
+			return this.createClusterEndpoint(clusterEndpoint, clusterId);
+		} catch (Exception e) {
+			this.client.logException(e);
+			this.tryKillCluster(clusterId);
+			throw new ClusterDeploymentException("Could not create Kubernetes cluster " + clusterId, e);
+		}
+	}
+
+	/**
+	 * Try to kill cluster without throw exception.
+	 */
+	private void tryKillCluster(String clusterId) {
+		try {
+			this.killCluster(clusterId);
+		} catch (Exception e) {
+			this.client.logException(e);
+		}
+	}
+
+	@Override
+	public void killCluster(String clusterId) throws FlinkException {
+		try {
+			this.client.stopAndCleanupCluster(clusterId);
+		} catch (Exception e) {
+			this.client.logException(e);
+			throw new FlinkException("Could not create Kubernetes cluster " + clusterId);
```

Comment:
Is this a typo for "Could not create Kubernetes cluster" in `killCluster` method.